### PR TITLE
Addon-docs: Support jsdoc params to describe function signature

### PR DESCRIPTION
**Pull request by patricklafrance**
Issue: https://github.com/storybookjs/storybook/issues/8532

## What I did

This PR add JSDoc support to describe the signature of a function in the docs addon props table when using PropTypes.

When a prop type is defined as a function, if the optional JSDoc is provided, the type of the prop will be a function signature infered from the JSDoc @param and @returns tags instead of being hardcoded to "func".

The support for @param and @returns follow the JSDoc specs: 

https://jsdoc.app/tags-param.html
https://jsdoc.app/tags-returns.html

This PR is a base for other issues and enhancement that will be tackled soon.

There is a few considerations that you should be aware of...

**Limitations:**

A property description cannot start with a @.

This means that for the property:

```
  /**
   * @onClick description
   */
  onClick: PropTypes.func
```

The description would be empty.

When @ is in the body of the description it's fine.

```
  /**
   * onClick @description@
   */
  onClick: PropTypes.func
```

Would display: onClick @description@

**JSDoc parser:**

To parse the JSDoc, I found 2 popular packages: https://www.npmjs.com/package/doctrine and https://www.npmjs.com/package/comment-parser

doctrine is obviously the most popular one with 14 millions weekly downloads. Sadly, it has been deprecated a few months ago.

The ESLint team decided to stop supporting rules to validate JSDoc format.

comment-parser is an interesting option but it doesn't work in a web environment, it has a dependency on node fs.

I decided to use doctrine even if it's deprecated since it's still widely use and is stable. I don't think we will end up supporting something else than @param, @returns and @ignore which works fine with doctrine. I dont see the JSDoc specs change for those tag in a near future.

**Parsing the JSDoc:**

To ensure good perfomance, the JSDoc tags should probably be parsed at compile time with a babel plugin. To get something out quickly I parse them from the Props React component.

Performance seems acceptable for now. 3 tables of 50 props having JSDoc tags render instantly in dev. A test is available in the cra-ts-kitchen-sink.

**What's next?**

- Display the @params and @returns under the function signature when descriptions are available. This should be done PropTypes and TS
- Add support for @ignore. View issue https://github.com/storybookjs/storybook/issues/8538

## How to test

- Jest tests are avaiable: `yarn jest --testPathPattern=type-system-handlers --watch`
- Performance tests are available in `cra-ts-kitchen-sink` under the story "docgen/Props types/jsdoc-perfo"
- 43 rendering cases are available `cra-ts-kitchen-sink` under the story "docgen/Props types/jsdoc"

### CODE REVIEW

**Comment by patricklafrance** [comment]
I am not sure how I ended up with all those changes to the `cra-ts-kichen-sink` @shilman .

I only changed the stories.

**Comment by shilman** [comment]
@patricklafrance This is amazing work. 🥇 

The changes are there because you started on my branch, which already upgraded `cra-ts-kitchen-sink` prior to your changes.

I'll review it all properly today or this weekend, but at first glance it looks superb!!!

File: addons/docs/package.json [hunk 1]
```diff
@@ -59,6 +60,9 @@
     "ts-dedent": "^1.1.0"
   },
   "devDependencies": {
+    "@types/doctrine": "^0.0.3",
+    "@types/enzyme": "^3.10.3",
```
Comment (shilman) [inline addons/docs/package.json]: Is this needed?

File: addons/docs/src/frameworks/react/extractProps.ts [hunk 2]
```diff
@@ -70,9 +33,8 @@ export const getPropDefs: PropDefGetter = (type, section) => {
       processedType = type.type().type;
     }
   }
-  return hasDocgen(processedType)
-    ? propsFromDocgen(processedType, section)
-    : propsFromPropTypes(processedType, section);
+
+  return extractPropsFromDocgen(processedType, section);
```
Comment (shilman) [inline addons/docs/src/frameworks/react/extractProps.ts]: Is `propsFromPropTypes` unused? Is this a breaking change?

File: addons/docs/src/lib/type-system-handlers.ts [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+};
+
+export const TypeSystemHandlers: Record<string, TypeSystemHandler> = {
+  [TypeSystem.Flow]: flowHandler,
+  [TypeSystem.TypeScript]: tsHandler,
+  [TypeSystem.PropTypes]: propTypesHandler,
+  [TypeSystem.Unknown]: unknownHandler,
+};
+
+export const getPropTypeSystem = (docgenInfo: DocgenInfo): string => {
```
Comment (shilman) [inline addons/docs/src/lib/type-system-handlers.ts]: This is wonderful ❤️

File: addons/docs/src/frameworks/react/extractProps.ts (hunk 2 above)
Comment (shilman) [reply addons/docs/src/frameworks/react/extractProps.ts]: Nvm, I see it below. Is this tested?

Comment (shilman) [reply addons/docs/src/frameworks/react/extractProps.ts]: Nvm, I see it is! 😻

**Review by shilman — APPROVED** [review]
This is tremendous. Totally blown away!!! ❤️ 

Minor comments only...

File: addons/docs/package.json (hunk 1 above)
Comment (patricklafrance) [reply addons/docs/package.json]: Without those types I couldn't compile the app witht `yarn dev`

**Pull request merged**